### ISSUE TITLE
Training dataset

### ISSUE THREAD

**Issue by sahilshukla3003** [Xiaojiu-z/Stable-Hair#18]

**Comment by sahilshukla3003** [comment]
how to convert the source image into triplet data form , can you please provide the code for that... , it will help a lot.

**Comment by sahilshukla3003** [comment]
I have tried step 2 using source image , inverted mask with inpainting model and some basic prompts , but not getting good results.

import PIL
import torch
from PIL import Image
from diffusers import StableDiffusionInpaintPipeline
import numpy as np

# Function to load an image from a local file path
def load_image(image_path):
    return Image.open(image_path).convert("RGB")

# Function to invert the mask
def invert_mask(mask):
    # Convert the mask to a numpy array and invert it (0 becomes 255 and vice versa)
    mask_array = np.array(mask)
    inverted_mask_array = 255 - mask_array
    return Image.fromarray(inverted_mask_array)

# Local paths of the original image and its mask (hairstyle mask in this case)
img_path = "01.jpg"
mask_path = "01_mask.png"

# Load and resize the images to the required size
init_image = load_image(img_path).resize((512, 512))
mask_image = load_image(mask_path).resize((512, 512))

# Invert the mask to make the non-hairstyle portions to be inpainted
inverted_mask_image = invert_mask(mask_image)

# Load the Stable Diffusion inpainting pipeline
pipe = StableDiffusionInpaintPipeline.from_pretrained(
    "runwayml/stable-diffusion-inpainting",
    torch_dtype=torch.float16,
)
pipe = pipe.to("cuda")

# Set the prompt to alter the non-hairstyle portions of the image (background/identity)
#prompt = "A photo of person standing in a studio"
prompt = "A portrait of a person with natural face and wearing black t-shirt, high resolution."

# Generate the new image by inpainting the non-hairstyle portions
image = pipe(prompt=prompt, image=init_image, mask_image=inverted_mask_image).images[0]

print("Done ------------------------------")

# Save or display the resulting image
image.save("inpainted_output.png")
# image.show()

![Screenshot from 2024-11-07 19-30-14](https://github.com/user-attachments/assets/f862acd4-7009-4575-8a83-55971e270bfe)


what is the issue..?

**Comment by Xiaojiu-z** [comment]
> I have tried step 2 using source image , inverted mask with inpainting model and some basic prompts , but not getting good results.
> 
> import PIL import torch from PIL import Image from diffusers import StableDiffusionInpaintPipeline import numpy as np
> 
> def load_image(image_path): return Image.open(image_path).convert("RGB")
> 
> def invert_mask(mask): # Convert the mask to a numpy array and invert it (0 becomes 255 and vice versa) mask_array = np.array(mask) inverted_mask_array = 255 - mask_array return Image.fromarray(inverted_mask_array)
> 
> img_path = "01.jpg" mask_path = "01_mask.png"
> 
> init_image = load_image(img_path).resize((512, 512)) mask_image = load_image(mask_path).resize((512, 512))
> 
> inverted_mask_image = invert_mask(mask_image)
> 
> pipe = StableDiffusionInpaintPipeline.from_pretrained( "runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, ) pipe = pipe.to("cuda")
> 
> #prompt = "A photo of person standing in a studio" prompt = "A portrait of a person with natural face and wearing black t-shirt, high resolution."
> 
> image = pipe(prompt=prompt, image=init_image, mask_image=inverted_mask_image).images[0]
> 
> print("Done ------------------------------")
> 
> image.save("inpainted_output.png")
> 
> ![Screenshot from 2024-11-07 19-30-14](https://private-user-images.githubusercontent.com/124785012/383991905-f862acd4-7009-4575-8a83-55971e270bfe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5OTMyNjgsIm5iZiI6MTczMDk5Mjk2OCwicGF0aCI6Ii8xMjQ3ODUwMTIvMzgzOTkxOTA1LWY4NjJhY2Q0LTcwMDktNDU3NS04YTgzLTU1OTcxZTI3MGJmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwN1QxNTIyNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzJiOTNmODRiYmY4MDhiNWViNWZiOTlhM2U1NmU2M2QzMWYwOWExZmM3NTQzN2U1N2ExM2E3NGY5NmI0ODg3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PH500aeqyL0iOvcNpbNEIlKBPT58RLoAthyF1Kc40BQ)
> 
> what is the issue..?

Actually, It's enough to change the identity and background, and you also need to scale the face and mask to simulate the small face condition. And the most important thing is using my bold converter to get bald images.

**Comment by sahilshukla3003** [comment]
> > I have tried step 2 using source image , inverted mask with inpainting model and some basic prompts , but not getting good results.
> > import PIL import torch from PIL import Image from diffusers import StableDiffusionInpaintPipeline import numpy as np
> > def load_image(image_path): return Image.open(image_path).convert("RGB")
> > def invert_mask(mask): # Convert the mask to a numpy array and invert it (0 becomes 255 and vice versa) mask_array = np.array(mask) inverted_mask_array = 255 - mask_array return Image.fromarray(inverted_mask_array)
> > img_path = "01.jpg" mask_path = "01_mask.png"
> > init_image = load_image(img_path).resize((512, 512)) mask_image = load_image(mask_path).resize((512, 512))
> > inverted_mask_image = invert_mask(mask_image)
> > pipe = StableDiffusionInpaintPipeline.from_pretrained( "runwayml/stable-diffusion-inpainting", torch_dtype=torch.float16, ) pipe = pipe.to("cuda")
> > #prompt = "A photo of person standing in a studio" prompt = "A portrait of a person with natural face and wearing black t-shirt, high resolution."
> > image = pipe(prompt=prompt, image=init_image, mask_image=inverted_mask_image).images[0]
> > print("Done ------------------------------")
> > image.save("inpainted_output.png")
> > ![Screenshot from 2024-11-07 19-30-14](https://private-user-images.githubusercontent.com/124785012/383991905-f862acd4-7009-4575-8a83-55971e270bfe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA5OTMyNjgsIm5iZiI6MTczMDk5Mjk2OCwicGF0aCI6Ii8xMjQ3ODUwMTIvMzgzOTkxOTA1LWY4NjJhY2Q0LTcwMDktNDU3NS04YTgzLTU1OTcxZTI3MGJmZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTA3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwN1QxNTIyNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hNzJiOTNmODRiYmY4MDhiNWViNWZiOTlhM2U1NmU2M2QzMWYwOWExZmM3NTQzN2U1N2ExM2E3NGY5NmI0ODg3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PH500aeqyL0iOvcNpbNEIlKBPT58RLoAthyF1Kc40BQ)
> > what is the issue..?
> 
> Actually, It's enough to change the identity and background, and you also need to scale the face and mask to simulate the small face condition. And the most important thing is using my bold converter to get bald images.

I understand the first step that is a bald converter for generating bald_proxy_images , and step two is as mentioned above to generate reference images and this way i will complete the triplet data generation which is required to train the stage 2 . Am i right..?

**Comment by sahilshukla3003** [comment]
And in step two you mention that scaling the face and mask is essential , so scaling will solve my problem of getting distorted images by inpainting

**Comment by Xiaojiu-z** [comment]
> And in step two you mention that scaling the face and mask is essential , so scaling will solve my problem of getting distorted images by inpainting....? or how will i generate the good reference images.

![image](https://github.com/user-attachments/assets/7424a0a0-e324-4571-84ce-22a958ca852d)
These two images are good examples, because the reference image's identity and background are changed, and it's good enough to be a reference image. Scaling is a great way to simulate the small face reference condition, It's up to you to decide whether to use it.

**Comment by sahilshukla3003** [comment]
Thanks for your reply , I have few more questions as I wanted to train stage 2 :
1) In train_stage2.sh 

    --refer_column="reference" \
    --source_column="source" \
    --target_column="target" \
    
    refer_column = generated reference images
    source_column = original source images
    target_column = bald proxy images

Is this correct ...?

**Comment by sahilshukla3003** [comment]
> > Thanks for your reply , I have few more questions as I wanted to train stage 2 :
> > 
> > 1. In train_stage2.sh
> >    --refer_column="reference"
> >    --source_column="source"
> >    --target_column="target" 
> >    refer_column = generated reference images
> >    source_column = original source images
> >    target_column = bald proxy images
> > 
> > Is this correct ...?
> 
> refer_column = generated reference images source_column = bald proxy images target_column = original source images

Thanks for your quick reply , looking forward to train on long hairs dataset!
